src/dcd/server/autocomplete/ufcs.d

```diff
@@ -14,11 +14,15 @@ import dparse.lexer : tok;
 import std.regex;
 import containers.hashset : HashSet;
 
-void lookupUFCS(Scope* completionScope, DSymbol* beforeDotSymbol, size_t cursorPosition, ref AutocompleteResponse response)
+bool lookupUFCS(Scope* completionScope, DSymbol* beforeDotSymbol, size_t cursorPosition, ref AutocompleteResponse response)
 {
-    // UFCS completion
+    if (beforeDotSymbol.isInvalidForUFCSCompletion)
+        return false;
     DSymbol*[] ufcsSymbols = getSymbolsForUFCS(completionScope, beforeDotSymbol, cursorPosition);
+    if (ufcsSymbols.empty)
+        return false;
     response.completions ~= map!(s => createCompletionForUFCS(s))(ufcsSymbols).array;
```

Comment:
The `.array` should not be needed here.

The simplest replacement I can think of is

```d
response.completions.reserve(response.completions.length + ufcsSymbols.length);
foreach (s; ufcsSymbols)
    response.completions ~= s.createCompletionForUFCS;
```

. Ideally the compiler would understand how to do

```d
response.completions ~= map!(s => createCompletionForUFCS(s))(ufcsSymbols);
```

or at least make the allocation in `.array` in

```d
response.completions ~= map!(s => createCompletionForUFCS(s))(ufcsSymbols).array;
```

use scoped allocation having deterministic destruction. For reference see https://dlang.org/library/object/reserve.html.